src/lib.rs

```diff
@@ -162,3 +208,371 @@ impl LevelIndex {
+            samples.push(SampleInformation::from_bytes(&bytes[offset..offset + 16]));
+            offset += 16;
+        }
+        assert_eq!(offset as u32, descriptor_block_size);
+
+        Self {
+            vendor_id,
+            descriptor_type,
+            version_number,
+            descriptor_block_size,
+            color_model: ColorModel::from(model),
+            color_primaries: ColorPrimaries::from(primaries),
+            transfer_function: TransferFunction::from(transfer),
+            flags: DataFormatFlags::from_bits_truncate(flags),
+            texel_block_dimensions,
+            bytes_planes,
+            samples,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SampleInformation {
+    pub bit_offset: u32,                                //: 16;
+    pub bit_length: u32,                                //: 8;
+    pub channel_type: u32,                              //: 4;
+    pub channel_type_qualifiers: ChannelTypeQualifiers, //: 4;
+    pub sample_positions: [u32; 4],                     //: 8 x 4;
+    pub lower: u32,                                     //;
+    pub upper: u32,                                     //;
+}
+
+impl SampleInformation {
+    const LENGTH: usize = 16;
+
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        let mut offset = 0;
+
+        let v = bytes_to_u32(bytes, &mut offset);
+        let bit_offset = shift_and_mask_lower(0, 16, v);
+        let bit_length = shift_and_mask_lower(16, 8, v) + 1;
+        let channel_type = shift_and_mask_lower(24, 4, v);
+        let channel_type_qualifiers = ChannelTypeQualifiers::from_bits_truncate(shift_and_mask_lower(24, 4, v));
+
+        let v = bytes_to_u32(bytes, &mut offset);
+        let sample_positions = [
+            shift_and_mask_lower(0, 8, v),
+            shift_and_mask_lower(8, 8, v),
+            shift_and_mask_lower(16, 8, v),
+            shift_and_mask_lower(24, 8, v),
+        ];
+        let lower = bytes_to_u32(bytes, &mut offset);
+        let upper = bytes_to_u32(bytes, &mut offset);
+
+        assert_eq!(offset, 16);
+
+        Self {
+            bit_offset,
+            bit_length,
+            channel_type,
+            channel_type_qualifiers,
+            sample_positions,
+            lower,
+            upper,
+        }
+    }
+
+    pub fn is_exponent(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::EXPONENT)
+    }
+
+    pub fn is_float(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::FLOAT)
+    }
+
+    pub fn is_linear(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::LINEAR)
+    }
+
+    pub fn is_signed(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::SIGNED)
+    }
+
+    pub fn is_norm(&self) -> bool {
+        self.is_float() && self.upper == F32_1_AS_U32
+    }
+}
+
+#[derive(Debug)]
+pub enum ColorModel {
```

Comment:
Extensible enums should use the `pseudo_enum` macro from enum.rs (and perhaps be defined there as well). This allows new values to be added without breaking compatibility.

Reply:
Interesting, I’ll take a look at that.

Reply:
That is the intended pattern, thanks!